### PR TITLE
port of betaflight#9190: restructure crossfire frame processing and remote control channel handling to prevent erroneous dropped frames

### PR DESCRIPTION
**Pull request by gretel**
port of https://github.com/betaflight/betaflight/pull/9190/files restructure crossfire frame processing and remote control channel handling to prevent erroneous dropped frames.
tested on `FURYF4OSD` using crossfire firmware 4.05 👍

### CODE REVIEW

File: src/test/unit/rx_crsf_unittest.cc [hunk 1]
```diff
@@ -176,6 +177,7 @@ TEST(CrossFireTest, TestCrsfFrameStatusUnpacking)
     const uint8_t crc = crsfFrameCRC();
     crsfFrame.frame.payload[CRSF_FRAME_RC_CHANNELS_PAYLOAD_SIZE] = crc;
 
+    memcpy(&crsfChannelDataFrame, &crsfFrame, sizeof(crsfFrame));
```
Comment (nerdCopter) [inline src/test/unit/rx_crsf_unittest.cc:180]: Codacy found an issue: [Does not check for buffer overflows when copying to destination (CWE-120).  Make sure destination can always hold the source data. ](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=6310371)

**Comment by nerdCopter** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=6310371&bid=20519674)

File: src/main/rx/crsf.c [hunk 2]
```diff
@@ -184,46 +185,52 @@ STATIC_UNIT_TESTED void crsfDataReceive(uint16_t c, void *data)
...
-                    switch (crsfFrame.frame.type)
-                    {
+            const uint8_t crc = crsfFrameCRC();
+            if (crc == crsfFrame.bytes[fullFrameLength - 1]) {
+                switch (crsfFrame.frame.type)
+                {
+                    case CRSF_FRAMETYPE_RC_CHANNELS_PACKED:
+                        if (crsfFrame.frame.deviceAddress == CRSF_ADDRESS_FLIGHT_CONTROLLER) {
+                            crsfFrameDone = true;
+                            memcpy(&crsfChannelDataFrame, &crsfFrame, sizeof(crsfFrame));
```
Comment (nerdCopter) [inline src/main/rx/crsf.c:197]: Codacy found an issue: [Does not check for buffer overflows when copying to destination (CWE-120).  Make sure destination can always hold the source data. ](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=6310371)

Comment (gretel) [reply src/main/rx/crsf.c:197]: interesting.

**Comment by Quick-Flash** [comment]
awesome, codacy see's some potential issues, but i tried porting this a while ago to no avail. Thanks for figuring it out :) lets go!

**Comment by gretel** [comment]
i'm happy this build pipeline is working fine.

**Comment by gretel** [comment]
think it's safe enough cause of pointers and `sizeof()` but i'm not sure. it's like this in bf still.

**Review by Quick-Flash — DISMISSED** [review]
Looks good to me. Thanks for this work :+1:

**Comment by nerdCopter** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=6310371&bid=20519674)

**Comment by gretel** [comment]
how can we get this merged?

**Review by Quick-Flash — APPROVED** [review]
Thanks for porting this over

**Pull request merged**